cannon/mipsevm/tests/evm_common_test.go

```diff
@@ -231,6 +231,75 @@ func TestEVMSingleStep_Operators(t *testing.T) {
+		{name: "lw", opcode: uint32(0x23), memVal: uint32(0x12_23_45_67), expectRes: uint32(0x12_23_45_67)},                                                                        // lw $t0, 4($t1)
+		{name: "lbu", opcode: uint32(0x24), memVal: uint32(0x12_23_00_00), expectRes: uint32(0x12)},                                                                                // lbu $t0, 4($t1)
+		{name: "lhu", opcode: uint32(0x25), memVal: uint32(0x12_23_00_00), expectRes: uint32(0x12_23)},                                                                             // lhu $t0, 4($t1)
+		{name: "lwl", opcode: uint32(0x22), rt: uint32(0xaa_bb_cc_dd), memVal: uint32(0x12_34_56_78), expectRes: uint32(0x12_34_56_78)},                                            // lwl $t0, 4($t1)
+		{name: "lwl unaligned address", opcode: uint32(0x22), rt: uint32(0xaa_bb_cc_dd), isUnAligned: true, memVal: uint32(0x12_34_56_78), expectRes: uint32(0x34_56_78_dd)},       // lwl $t0, 5($t1)
+		{name: "lwr", opcode: uint32(0x26), rt: uint32(0xaa_bb_cc_dd), memVal: uint32(0x12_34_56_78), expectRes: uint32(0xaa_bb_cc_12)},                                            // lwr $t0, 4($t1)
+		{name: "lwr unaligned address", opcode: uint32(0x26), rt: uint32(0xaa_bb_cc_dd), isUnAligned: true, memVal: uint32(0x12_34_56_78), expectRes: uint32(0xaa_bb_12_34)},       // lwr $t0, 5($t1)
+		{name: "sb", opcode: uint32(0x28), rt: uint32(0xaa_bb_cc_dd), expectRes: uint32(0xaa_bb_cc_dd), expectMemVal: uint32(0xdd_00_00_00)},                                       // sb $t0, 4($t1)
+		{name: "sh", opcode: uint32(0x29), rt: uint32(0xaa_bb_cc_dd), expectRes: uint32(0xaa_bb_cc_dd), expectMemVal: uint32(0xcc_dd_00_00)},                                       // sh $t0, 4($t1)
+		{name: "swl", opcode: uint32(0x2a), rt: uint32(0xaa_bb_cc_dd), expectRes: uint32(0xaa_bb_cc_dd), expectMemVal: uint32(0xaa_bb_cc_dd)},                                      //  swl $t0, 4($t1)
+		{name: "sw", opcode: uint32(0x2b), rt: uint32(0xaa_bb_cc_dd), expectRes: uint32(0xaa_bb_cc_dd), expectMemVal: uint32(0xaa_bb_cc_dd)},                                       // sw $t0, 4($t1)
+		{name: "swr unaligned address", opcode: uint32(0x2e), rt: uint32(0xaa_bb_cc_dd), isUnAligned: true, expectRes: uint32(0xaa_bb_cc_dd), expectMemVal: uint32(0xcc_dd_00_00)}, // swr $t0, 5($t1)
+	}
+
+	var t1 uint32 = 0x100
+	for _, v := range versions {
+		for i, tt := range cases {
+			testName := fmt.Sprintf("%v (%v)", tt.name, v.Name)
+			t.Run(testName, func(t *testing.T) {
+				goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), testutil.WithRandomization(int64(i)), testutil.WithPC(0), testutil.WithNextPC(4))
+				state := goVm.GetState()
+				var insn uint32
+				imm := uint32(0x4)
+				if tt.isUnAligned {
+					imm = uint32(0x5)
+				}
+
+				insn = tt.opcode<<26 | uint32(9)<<21 | uint32(8)<<16 | imm
```

Comment:
(nit) For clarity, would be helpful to replace the [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants) here with variables, something like:
```suggestion
				insn = tt.opcode<<26 | uint32(baseReg)<<21 | uint32(rtReg)<<16 | imm
```